compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java

```diff
@@ -5125,42 +5128,46 @@ onFailErrorVariableSymbol.name.value, onFailErrorVariableSymbol.type, rewrite(fa
-            onFailBody.scope.define(onFailErrorVariableSymbol.name, onFailErrorVariableSymbol);
-            onFailBody.stmts.add(0, errorVarDef);
-        }
-
-        int currentOnFailIndex = this.enclosingOnFailClause.indexOf(this.onFailClause);
-        int enclosingOnFailIndex = currentOnFailIndex <= 0 ? this.enclosingOnFailClause.size() - 1
-                : (currentOnFailIndex - 1);
-        this.onFailClause = this.enclosingOnFailClause.get(enclosingOnFailIndex);
-        onFailBody = rewrite(onFailBody, env);
-        BLangFail failToEndBlock = new BLangFail();
-        if (onFailClause.isInternal && fail.exprStmt != null) {
-            if (fail.exprStmt instanceof BLangPanic) {
-                setPanicErrorToTrue(onFailBody, onFailClause);
-            } else {
-                onFailBody.stmts.add((BLangStatement) fail.exprStmt);
+        BLangBlockStmt onFailBody = blockStmtByFailNode.get(fail);
+        if (onFailBody == null) {
+            onFailBody = ASTBuilderUtil.createBlockStmt(onFailClause.pos);
+            onFailBody.stmts.addAll(onFailClause.body.stmts);
+            onFailBody.scope = onFailClause.body.scope;
+            onFailBody.mapSymbol = onFailClause.body.mapSymbol;
+            onFailBody.failureBreakMode = onFailClause.body.failureBreakMode;
+            onFailBody.isFromFailNode = true;
+            VariableDefinitionNode onFailVarDefNode = onFailClause.variableDefinitionNode;
+
+            if (onFailVarDefNode != null) {
+                BVarSymbol onFailErrorVariableSymbol =
+                        ((BLangSimpleVariableDef) onFailVarDefNode).var.symbol;
+                BLangSimpleVariable errorVar = ASTBuilderUtil.createVariable(onFailErrorVariableSymbol.pos,
+                        onFailErrorVariableSymbol.name.value, onFailErrorVariableSymbol.type, rewrite(fail.expr, env),
+                        onFailErrorVariableSymbol);
+                BLangSimpleVariableDef errorVarDef = ASTBuilderUtil.createVariableDef(onFailClause.pos, errorVar);
+                onFailBody.scope.define(onFailErrorVariableSymbol.name, onFailErrorVariableSymbol);
+                onFailBody.stmts.add(0, errorVarDef);
+            }
+
+            int currentOnFailIndex = this.enclosingOnFailClause.indexOf(this.onFailClause);
+            int enclosingOnFailIndex = currentOnFailIndex <= 0 ? this.enclosingOnFailClause.size() - 1
+                    : (currentOnFailIndex - 1);
+            this.onFailClause = this.enclosingOnFailClause.get(enclosingOnFailIndex);
+            onFailBody = rewrite(onFailBody, env);
+            BLangFail failToEndBlock = new BLangFail();
+            if (onFailClause.isInternal && fail.exprStmt != null) {
+                if (fail.exprStmt instanceof BLangPanic) {
+                    setPanicErrorToTrue(onFailBody, onFailClause);
+                } else {
+                    onFailBody.stmts.add((BLangStatement) fail.exprStmt);
```

Comment:
Shall we add tests to cover the missing lines

Reply:
Those lines are not newly added lines and it seems those lines are not covered by any ballerina-lang tests and even not used by the `module-ballerinai-transaction` module. Created a separate task to address this https://github.com/ballerina-platform/ballerina-lang/issues/39017.